packages/gatsby/src/schema/infer-graphql-input-fields.js

```diff
@@ -205,6 +205,7 @@ export function inferInputObjectStructureFromNodes({
       prefix: `${prefix}${_.upperFirst(key)}`,
     })
 
+    if (typeof key !== 'string') return
```

Comment:
This isn't the right fix — source/transformer plugins need to ensure that they're not passing in bad data. Add a console.log here to see what bad data gatsby-source-mongodb is passing in and fix it there.